Change TrtConversionParams to class from NamedTuple and export it

pooyadavoodi opened the pull request:
- Changes NamedTuple to a class.
- Supports backward compatibility by providing _replace() method to emulate the behavior of _replace() in NamedTuple.
- Exports ConversionParams to API

Commit:
Change TrtConversionParams to class from NamedTuple

This change preserves backward compatibility of TF-TRT API.

aaroey commented:
@pooyadavoodi would you please fix the ubuntu sanity errors? Thanks.

Commit:
Remove unused import and do not use "is" for string comparison

pooyadavoodi commented:
> @pooyadavoodi would you please fix the ubuntu sanity errors? Thanks.

Done. Thanks @aaroey 

I always forget to run pylint before pushing PRs. I wish there was a way to automate this.

aaroey commented:
@pooyadavoodi this breaks tensorflow/tools/api/tests:api_compatibility_test, could you take a look?

Commit:
Update golden API for TF-TRT backward compatibility test

pooyadavoodi commented:
> @pooyadavoodi this breaks tensorflow/tools/api/tests:api_compatibility_test, could you take a look?

I believe I fixed the golden api in the last commit. I had to rebase onto the latest master because there was a keras api test failure.

pooyadavoodi commented:
@sanjoy Please let me know if I should do anything for this PR?

sanjoy commented:
> @sanjoy Please let me know if I should do anything for this PR?

Thanks for pinging!  Could be that folks who usually do the merge are OOO for holidays.

@gbaned what needs to happen here?

gbaned commented:
> > @sanjoy Please let me know if I should do anything for this PR?
> 
> Thanks for pinging! Could be that folks who usually do the merge are OOO for holidays.
> 
> @gbaned what needs to happen here?

@sanjoy  import/copybara check is failing. We can proceed further once it is successful. Thank you!

pooyadavoodi commented:
@gbaned I don't know anything about import/copybara. It seems that the failure has no log. How can I check what actually failed?

rthadur commented:
@pooyadavoodi here is the internal error , can you please chekc once 
`Traceback (most recent call last):
  File "<embedded stdlib>/unittest/case.py", line 59, in testPartExecutor
    yield
  File "<embedded stdlib>/unittest/case.py", line 605, in run
    testMethod()
  File "/google3/runfiles/google3/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 428, in testAPIBackwardsCompatibilityV2
    omit_golden_symbols_map=omit_golden_symbols_map)
  File "/google3/runfiles/google3/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 367, in _checkBackwardsCompatibility
    api_version=api_version)
  File "/google3/runfiles/google3/third_party/tensorflow/tools/api/tests/api_compatibility_test.py", line 287, in _AssertProtoDictEquals
    self.fail('%d differences found between API and golden.' % diff_count)
  File "/google3/runfiles/google3/third_party/py/absl/testing/absltest.py", line 1761, in fail
    return super(TestCase, self).fail(self._formatMessage(prefix, msg))
  File "<embedded stdlib>/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: 1 differences found between API and golden.
`

pooyadavoodi commented:
@rthadur @sanjoy  I can't repro this error.

This is what I am getting when running test `//tensorflow/tools/api/tests:api_compatibility_test`

```
INFO:tensorflow:No differences found between API and golden.
I0106 19:41:36.972908 140491009787712 api_compatibility_test.py:290] No differences found between API and golden.
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
INFO:tensorflow:No differences found between API and golden.
I0106 19:41:39.017542 140491009787712 api_compatibility_test.py:290] No differences found between API and golden.
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessage
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessage
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[ RUN      ] ApiCompatibilityTest.test_session
[  SKIPPED ] ApiCompatibilityTest.test_session
----------------------------------------------------------------------
Ran 7 tests in 8.121s

OK (skipped=1)
================================================================================
Target //tensorflow/tools/api/tests:api_compatibility_test up-to-date:
  bazel-bin/tensorflow/tools/api/tests/api_compatibility_test
INFO: Elapsed time: 1270.347s, Critical Path: 716.82s
INFO: 6652 processes: 6652 local.
INFO: Build completed successfully, 6655 total actions
//tensorflow/tools/api/tests:api_compatibility_test                      PASSED in 12.1s
  WARNING: //tensorflow/tools/api/tests:api_compatibility_test: Test execution time (12.1s excluding execution overhead) outside of range for MODERATE tests. Consider setting timeout="short" or size="small".

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 6655 total actions
```

rthadur commented:
@pooyadavoodi thank you , @sanjoy can you please suggest ?

sanjoy commented:
@annarev we have a case where `ApiCompatibilityTest.testAPIBackwardsCompatibilityV2` is failing in google3 after the import but [passing](https://github.com/tensorflow/tensorflow/pull/35198#issuecomment-571309341) in open source.  Any idea what's going on?

aaroey reviewed (changes requested):
Try to import the PR again.

sanjoy commented:
@alextp had this comment in the internal version of this CL:

> (for tf-api-owners)
> 
> 1. It looks like the default argument is now mutable. This is dangerous as assigning to it can have unexpected consequences to unrelated code. Preserving immutability is a good idea (you can use patterns like .withchange(change) methods which return copies of the tuple).
> 
> 2. It doesn't look like the changed type was exposed in the public API. If so, how are users supposed to construct it to pass as arguments to that function?

Commit:
Export ConversionParams an API

Commit:
Let _replace() return a copy of object instead of changing the object

The idea is to make DEFAULT_TRT_CONVERSION_PARAMS immutable.
Although this doesn't make it completely immutable but it avoids
changing the object by _replace(), which is the method everyone is
using on DEFAULT_TRT_CONVERSION_PARAMS when it was a NamedTuple.

pooyadavoodi commented:
> @alextp had this comment in the internal version of this CL:
> 
> > (for tf-api-owners)
> > 
> > 1. It looks like the default argument is now mutable. This is dangerous as assigning to it can have unexpected consequences to unrelated code. Preserving immutability is a good idea (you can use patterns like .withchange(change) methods which return copies of the tuple).
> > 2. It doesn't look like the changed type was exposed in the public API. If so, how are users supposed to construct it to pass as arguments to that function?

@sanjoy I made `_replace()` to return a copy of the object instead of changing it. Please let me know if that is sufficient.

I also exported `ConversionParams` which we previously missed in TF2.1.

sanjoy commented:
Passing the buck to @alextp -- Alex, can you PTAL if your comments in https://github.com/tensorflow/tensorflow/pull/35198#issuecomment-573832904 have been addressed?

Commit:
Delete argument allow_build_at_runtime

This was accidentally introduced. The argument will be added in
https://github.com/tensorflow/tensorflow/pull/34919

annarev commented:
> @annarev we have a case where `ApiCompatibilityTest.testAPIBackwardsCompatibilityV2` is failing in google3 after the import but [passing](https://github.com/tensorflow/tensorflow/pull/35198#issuecomment-571309341) in open source. Any idea what's going on?

Sorry just saw the comment. Is this still an issue?

alextp commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
````diff
@@ -865,7 +882,7 @@ class TrtGraphConverterV2(object):
 
      ```python
-     params = DEFAULT_TRT_CONVERSION_PARAMS._replace(
+     params = tf.experimental.tensorrt.ConversionParams(
````
this module doesn't "import tensorflow as tf" so it can't use tf.experimental as a symbol (I don't think this code even runs?)

alextp commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,88 @@ def supported_precision_modes():
+      if v and (k != "self"):
+        setattr(trt_conversion_params, k, v)
+    return trt_conversion_params
+
```
Can this class have a __str__ or __repr__ method so the API pbtxt file doesn't end up with python IDs in it?

pooyadavoodi replied:
@alextp I added `__repr__` that returns a string. 

I also updated the pbtxt file. 

I use dict style in the returned string which is different from what we had before for the NamedTuple (brace and colon vs. parenthesis and equal sign). Please let me know if this is fine.

Commit:
Add __repr__ method to TrtConversionParams to have class properties in API pbtxt

pooyadavoodi commented:
@sanjoy @alextp The test is failing on github but passing on my machine locally. Any idea why?

Results in my machine:
```
INFO:tensorflow:No differences found between API and golden.
I0117 18:57:16.801649 140141113898816 api_compatibility_test.py:290] No differences found between API and golden.
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
INFO:tensorflow:No differences found between API and golden.
I0117 18:57:19.938147 140141113898816 api_compatibility_test.py:290] No differences found between API and golden.
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessage
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessage
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[ RUN      ] ApiCompatibilityTest.test_session
[  SKIPPED ] ApiCompatibilityTest.test_session
----------------------------------------------------------------------
Ran 7 tests in 11.279s

OK (skipped=1)
```

sanjoy commented:
> > @annarev we have a case where `ApiCompatibilityTest.testAPIBackwardsCompatibilityV2` is failing in google3 after the import but [passing](https://github.com/tensorflow/tensorflow/pull/35198#issuecomment-571309341) in open source. Any idea what's going on?
> 
> Sorry just saw the comment. Is this still an issue?
> One problem could be if the goldens need to be updated both for TF 1.x and TF 2.x versions.

Hi @annarev, I think this is still an issue, see https://github.com/tensorflow/tensorflow/pull/35198#issuecomment-575753807 ?

pooyadavoodi commented:
Looks like when I run `bazel test tensorflow/tools/api/tests:api_compatibility_test`, it fails similar to github:
```
ERROR:tensorflow:1 differences found between API and golden.
E0122 01:03:45.536555 139929196537664 api_compatibility_test.py:264] 1 differences found between API and golden.
Issue 1 : New object tensorflow.experimental.tensorrt.ConversionParams found (added).
[  FAILED  ] ApiCompatibilityTest.testAPIBackwardsCompatibility
...
ERROR:tensorflow:1 differences found between API and golden.
E0122 01:03:54.671969 139929196537664 api_compatibility_test.py:264] 1 differences found between API and golden.
Issue 1 : New object tensorflow.experimental.tensorrt.ConversionParams found (added).
[  FAILED  ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessage
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessage
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[ RUN      ] ApiCompatibilityTest.test_session
[  SKIPPED ] ApiCompatibilityTest.test_session
======================================================================
FAIL: testAPIBackwardsCompatibility (__main__.ApiCompatibilityTest)
testAPIBackwardsCompatibility (__main__.ApiCompatibilityTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/233a431e54f1cbffefcb6321279dfb61/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/
tests/api_compatibility_test.py", line 390, in testAPIBackwardsCompatibility
    omit_golden_symbols_map=omit_golden_symbols_map)
  File "/root/.cache/bazel/_bazel_root/233a431e54f1cbffefcb6321279dfb61/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/
tests/api_compatibility_test.py", line 367, in _checkBackwardsCompatibility
    api_version=api_version)
  File "/root/.cache/bazel/_bazel_root/233a431e54f1cbffefcb6321279dfb61/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/
tests/api_compatibility_test.py", line 287, in _AssertProtoDictEquals
    self.fail('%d differences found between API and golden.' % diff_count)
  File "/root/.cache/bazel/_bazel_root/233a431e54f1cbffefcb6321279dfb61/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/absl_py/absl/testing/absltest.py", l
ine 1655, in fail
    return super(TestCase, self).fail(self._formatMessage(prefix, msg))
AssertionError: 1 differences found between API and golden.
```

But when I run `bazel test tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True`, although the log reports the same failure, but it reports the final result as OK.

```
INFO:tensorflow:No differences found between API and golden.
I0117 18:57:16.801649 140141113898816 api_compatibility_test.py:290] No differences found between API and golden.
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
INFO:tensorflow:No differences found between API and golden.
I0117 18:57:19.938147 140141113898816 api_compatibility_test.py:290] No differences found between API and golden.
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessage
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessage
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[ RUN      ] ApiCompatibilityTest.test_session
[  SKIPPED ] ApiCompatibilityTest.test_session
----------------------------------------------------------------------
Ran 7 tests in 11.279s

OK (skipped=1)
```

- Issue 1) When I run with `-- --update_goldens True`, it shows OK at the end although there are failures in the log.
- Issue 2) When I run with `-- --update_goldens True`, it doesn't update the golden.
- Question 1) Should I update the golden manually, given that the script doesn't work?
- Question 2) Should the manual update be adding a new pbtxt file given that we have exported a new object called `ConversionParams`?
- Question 3) What should be the name of the new pbtxt file for `ConversionParams`. Perhaps: 
`tensorflow/tools/api/golden/v2/tensorflow.experimental.tensorrt.-conversion-params.pbtxt`?

Commit:
Add golden pbtxt file for TensorRT ConversionParams

pooyadavoodi commented:
I added a new empty file manually, and then `-- --update_goldens True` updated the file.
It would be great to document the requirement of adding files manually if it's not already documented.

sanjoy commented:
> Looks like when I run `bazel test tensorflow/tools/api/tests:api_compatibility_test`, it fails similar to github:

Did you figure out why it wasn't failing in https://github.com/tensorflow/tensorflow/pull/35198#issuecomment-575753807?

Other than that, I'm waiting for @alextp here.

alextp commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -951,7 +971,7 @@ def __init__(self,
                input_saved_model_tags=None,
                input_saved_model_signature_key=None,
-               conversion_params=DEFAULT_TRT_CONVERSION_PARAMS):
+               conversion_params=TrtConversionParams()):
```
This still uses a mutable object as the default value of the function. That's quite dangerous as this object can be inadvertently overwritten.

tf-api-owners prefer keeping TrtConversionParams immutable and using a replace method or, if you do want it mutable, pass None here and create a default instance in the body of the function

See https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html for a reference on why mutable default arguments are dangerous.

pooyadavoodi replied:
> This still uses a mutable object as the default value of the function. That's quite dangerous as this object can be inadvertently overwritten.
> 
> tf-api-owners prefer keeping TrtConversionParams immutable and using a replace method or, if you do want it mutable, pass None here and create a default instance in the body of the function
> 
> See https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html for a reference on why mutable default arguments are dangerous.

@alextp Thanks for the link. I made `TrtConversionParams` immutable by overriding `__setattr__`. Please let me know if that's not enough. 

We can't change the default value to None due to backward compatibility.
We don't really need `TrtConversionParams` to be mutable.

sanjoy replied:
> We can't change the default value to None due to backward compatibility.

I may be missing some Python subtleties, but can you default it to `None` and then within the function do

```
if conversion_params is None:
  conversion_params = TrtConversionParams()
```
?

pooyadavoodi replied:
@sanjoy Thanks, that would actually work. I fixed that.

Commit:
Make TrtConversionParams immutable

alextp commented:
Backward compatibility doesn't require that we cannot change the default
value, just that the new default value has the same meaning as the old
default value. So Sanjoy's suggestion is what needs to happen her.e

On Thu, Jan 23, 2020 at 5:32 PM Sanjoy Das <notifications@github.com> wrote:

> *@sanjoy* commented on this pull request.
> ------------------------------
>
> In tensorflow/python/compiler/tensorrt/trt_convert.py
> <https://github.com/tensorflow/tensorflow/pull/35198#discussion_r370439586>
> :
>
> > @@ -951,7 +971,7 @@ def __init__(self,
>                 input_saved_model_dir=None,
>                 input_saved_model_tags=None,
>                 input_saved_model_signature_key=None,
> -               conversion_params=DEFAULT_TRT_CONVERSION_PARAMS):
> +               conversion_params=TrtConversionParams()):
>
> We can't change the default value to None due to backward compatibility.
>
> I may be missing some Python subtleties, but can you default it to None
> and then within the function do
>
> if conversion_params is None:
>   conversion_params = TrtConversionParams()
>
> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/35198?email_source=notifications&email_token=AAABHRJEHMBJ22ZATNO3TTLQ7JAKRA5CNFSM4J37OG22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCS42WVI#discussion_r370439586>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRNF3KVVAWGFLFWP7LLQ7JAKRANCNFSM4J37OG2Q>
> .
>


-- 
 - Alex

Commit:
Make conversion_params immutable by using None as its default value

pooyadavoodi commented:
> > Looks like when I run `bazel test tensorflow/tools/api/tests:api_compatibility_test`, it fails similar to github:
> 
> Did you figure out why it wasn't failing in [#35198 (comment)](https://github.com/tensorflow/tensorflow/pull/35198#issuecomment-575753807)?
> 
> Other than that, I'm waiting for @alextp here.

I think those two issues still exist (I worked around them by adding the new pbtxt file manually and looking at the whole log instead of only final results for failure reports):

- When I run with `-- --update_goldens True`, it shows OK at the end although there are failures in the log.
- When I run with `-- --update_goldens True`, it doesn't update the golden.

alextp commented:
@pooyadavoodi you might be running it on the wrong directory (it assumes it'll be run from within the root directory)

Commit:
Update goldens api of TRT converter after changing params default value

sanjoy commented:
The windows build is failing: https://source.cloud.google.com/results/invocations/1eb40532-983f-47e5-b172-f91a7c7f842c/targets/%2F%2Ftensorflow%2Ftools%2Fci_build%2Fbuilds:gen_win_out/log

```
ERROR: T:/src/github/tensorflow/tensorflow/python/keras/api/BUILD:116:1: Executing genrule //tensorflow/python/keras/api:keras_python_api_gen_compat_v1 failed (Exit 1)
Traceback (most recent call last):
  File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\tools\api\generator\create_python_api.py", line 27, in <module>
    from tensorflow.python.tools.api.generator import doc_srcs
  File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\__init__.py", line 85, in <module>
    from tensorflow.python.ops.standard_ops import *
  File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\ops\standard_ops.py", line 115, in <module>
    from tensorflow.python.compiler.tensorrt import trt_convert_windows as trt
  File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\compiler\tensorrt\trt_convert_windows.py", line 30, in <module>
    class TrtConversionParams(object):
  File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\compiler\tensorrt\trt_convert_windows.py", line 35, in TrtConversionParams
    max_workspace_size_bytes=DEFAULT_TRT_MAX_WORKSPACE_SIZE_BYTES,
NameError: name 'DEFAULT_TRT_MAX_WORKSPACE_SIZE_BYTES' is not defined
Target //tensorflow/tools/pip_package:build_pip_package failed to build
ERROR: T:/src/github/tensorflow/tensorflow/python/tools/BUILD:98:1 Executing genrule //tensorflow/python/keras/api:keras_python_api_gen_compat_v1 failed (Exit 1)
INFO: Elapsed time: 1336.958s, Critical Path: 546.13s
INFO: 5660 processes: 3285 remote cache hit, 2375 local.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

Commit:
Add necessary variables for trt_convert in windows build

pooyadavoodi commented:
> The windows build is failing: https://source.cloud.google.com/results/invocations/1eb40532-983f-47e5-b172-f91a7c7f842c/targets/%2F%2Ftensorflow%2Ftools%2Fci_build%2Fbuilds:gen_win_out/log
> 
> ```
> ERROR: T:/src/github/tensorflow/tensorflow/python/keras/api/BUILD:116:1: Executing genrule //tensorflow/python/keras/api:keras_python_api_gen_compat_v1 failed (Exit 1)
> Traceback (most recent call last):
>   File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\tools\api\generator\create_python_api.py", line 27, in <module>
>     from tensorflow.python.tools.api.generator import doc_srcs
>   File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\__init__.py", line 85, in <module>
>     from tensorflow.python.ops.standard_ops import *
>   File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\ops\standard_ops.py", line 115, in <module>
>     from tensorflow.python.compiler.tensorrt import trt_convert_windows as trt
>   File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\compiler\tensorrt\trt_convert_windows.py", line 30, in <module>
>     class TrtConversionParams(object):
>   File "\\?\T:\tmp\Bazel.runfiles_aoh3x156\runfiles\org_tensorflow\tensorflow\python\compiler\tensorrt\trt_convert_windows.py", line 35, in TrtConversionParams
>     max_workspace_size_bytes=DEFAULT_TRT_MAX_WORKSPACE_SIZE_BYTES,
> NameError: name 'DEFAULT_TRT_MAX_WORKSPACE_SIZE_BYTES' is not defined
> Target //tensorflow/tools/pip_package:build_pip_package failed to build
> ERROR: T:/src/github/tensorflow/tensorflow/python/tools/BUILD:98:1 Executing genrule //tensorflow/python/keras/api:keras_python_api_gen_compat_v1 failed (Exit 1)
> INFO: Elapsed time: 1336.958s, Critical Path: 546.13s
> INFO: 5660 processes: 3285 remote cache hit, 2375 local.
> FAILED: Build did NOT complete successfully
> FAILED: Build did NOT complete successfully
> ```

I added the necessary definitions. 
Hopefully this fixes the issue. Unfortunately I can't build on Windows locally to test this.

sanjoy reviewed (approved):
lgtm

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+
+@tf_export("experimental.tensorrt.ConversionParams", v1=[])
+class TrtConversionParams(object):
+  """ A class to encapsulate parameters that are used for TF-TRT conversion."""
```
please fix this error 
`Extraneous space in one-line docstring`

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+    raise AttributeError(
+        "{} is immutable and thus cannot change its attributes".format(
+            self.__class__))
+
```
`Unused argument 'rewriter_config_template' [unused-argument]`

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+        "{} is immutable and thus cannot change its attributes".format(
+            self.__class__))
+
+  def _replace(self,
```
`Unused argument 'max_workspace_size_bytes' [unused-argument]`

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+            self.__class__))
+
+  def _replace(self,
+               rewriter_config_template=None,
```
`Unused argument 'precision_mode' [unused-argument]`

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+
+  def _replace(self,
+               rewriter_config_template=None,
+               max_workspace_size_bytes=None,
```
`Unused argument 'minimum_segment_size' [unused-argument]`

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+  def _replace(self,
+               rewriter_config_template=None,
+               max_workspace_size_bytes=None,
+               precision_mode=None,
```
`Unused argument 'is_dynamic_op' [unused-argument]`

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+               rewriter_config_template=None,
+               max_workspace_size_bytes=None,
+               precision_mode=None,
+               minimum_segment_size=None,
```
`Unused argument 'maximum_cached_engines' [unused-argument]`

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+               precision_mode=None,
+               minimum_segment_size=None,
+               is_dynamic_op=None,
+               maximum_cached_engines=None,
```
`Args section is missing, does not match arg list or is badly indented.

Expected: rewriter_config_template, max_workspace_size_bytes, precision_mode, minimum_segment_size, is_dynamic_op, maximum_cached_engines, use_calibration, max_batch_size

Found: 

Missing: rewriter_config_template, max_workspace_size_bytes, precision_mode, minimum_segment_size, is_dynamic_op, maximum_cached_engines, use_calibration, max_batch_size

Unexpected: `

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+               minimum_segment_size=None,
+               is_dynamic_op=None,
+               maximum_cached_engines=None,
+               use_calibration=None,
```
`Missing Returns section in docstring, or wrong choice of Returns/Yields `

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
+    set parameters:
+      DEFAULT_TRT_CONVERSION_PARAMS._replace(...)
+    """
+    conv_params_dict = {}
```
`Use 'is not' instead of '!=' when comparing to None`

rthadur commented on tensorflow/python/compiler/tensorrt/trt_convert_windows.py:
```diff
@@ -27,6 +27,72 @@
+
+@tf_export("experimental.tensorrt.ConversionParams", v1=[])
+class TrtConversionParams(object):
+  """ A class to encapsulate parameters that are used for TF-TRT conversion."""
```
`Extraneous space in one-line docstring `

sanjoy commented:
@rthadur I think it will be easier if I fixed these internally. Can you please import the CL and CC it to me?

sanjoy commented:
> @rthadur I think it will be easier if I fixed these internally. Can you please import the CL and CC it to me?

Sorry, looks like you already did that. :) Let me see if I can quickly fix these locally.

reedwm commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -111,70 +110,106 @@ def supported_precision_modes():
-    max_batch_size=1)
+
+@tf_export("experimental.tensorrt.ConversionParams", v1=[])
+class TrtConversionParams(object):
```
Drive by comment: Can you change this to subclass `namedtuple`, similar to what is done [here](https://github.com/tensorflow/tensorflow/blob/52eb7b53a484c088a30e19d380e8efb64ea423e3/tensorflow/python/ops/parsing_config.py#L57)? That way this still remains a namedtuple and you don't have to reimplement part of the namedtuple interface.

sanjoy replied:
Let's fix this in a follow-up commit.  CC @DEKHTIARJonathan

DEKHTIARJonathan replied:
@sanjoy let me know if you need me to do anything specific :)
Would be nice if that could be merged by TF 2.2 cut
